### ISSUE TITLE
Release 2.6.10 Condition with name="*" fails

### ISSUE THREAD

**Issue by tringate** [Program-O/Program-O#387]
The following testcases fail which test the condition using the name="*".

I am not totally clear on how the usage of the STAR is suppose to be used, but
I believe the test cases are testing it correctly.
```xml
<!-- test case 29 -->
<category><pattern>TESTCASE29</pattern><template>
<srai>CASE29 one</srai>
</template></category>

<category><pattern>CASE29 *</pattern><template>
<condition name="*" value="one">
*** PASSED ***
</condition>
</template></category>


<!-- test case 30 -->
<category><pattern>TESTCASE30</pattern><template>
<srai>CASE30 one</srai>
</template></category>

<category><pattern>CASE30 *</pattern><template>
<condition name="*">
<li value="one">*** PASSED ***</li>
<li>*** FAILED ***</li>
</condition>
</template></category>


<!-- test case 31 -->
<category><pattern>TESTCASE31</pattern><template>
<srai>CASE31 one</srai>
</template></category>

<category><pattern>CASE31 *</pattern><template>
<condition>
<li name="*" value="one">*** PASSED ***</li>
<li>*** FAILED ***</li>
</condition>
</template></category>
```

**Comment by OverlanAI** [comment]
Umm, could you simplify all that voodoo a bit?
I added that all in to see but i'm not getting the same result.
At best i can see here relates to another post about nested srai calls but i'm just a hacker.
Anyway i cant tell what the result SHOULD be.
TNX (i just have c++ foil hat not aiml)

**Comment by tringate** [comment]
This is currently a big that is turning out to be a bugger to get resolved.  All test cases either produce the results of *** PASSED *** or *** FAILED *** unless it is impossible to generate such a result.  Currently test cases 29, 30, 31 and 77, 78 are failing because of how the <star/> variable is being handled.  It may be one bug or two bugs.  The difficulty that has come into play while trying to fix these bugs is other bugs were found and have to be fixed first.

The other issue is that we do not have a true test case program that will send tests to a Program-O and analyze the results.  The current test cases are all written to be "self testing" which poses an issue of having to use a broken Program-O to test itself.  This requires a fully functional SRAI and STAR capability which is currently not working in the present dev version.  It is very close to having the most serious issues solved and then more obscure issues might be able to be discovered.  

The nested SRAI is also failing but we are not positive it isn't the same bug as with the STAR.  

Some tweaking of the voting system needed to be made to clear up issues we discovered while working on this STAR issue.

**Comment by patrickschur** [comment]
@tringate I didn't know you can use a star within a name attribute. Have you any documentation about this feature? I've never heard about it in the past.

**Comment by tringate** [comment]
patrickschur, I do not believe it is documented, at least I have never seen any such documentation.  The code for this extension has been in program-o for a fair amount of time.  Dave added it at some point and a single depth usage of it works fine.  He sent out an example a couple of months ago of how it worked with all the indexes.  You have to use the * and a number,  example: name="*3" would be the way to reference `<star index="3">`.  A single * by itself refers to the `<star/>`.  I had used it before in my very first version of my "genie" AIML but found a much better way to do things by writing new tags.  I decided to test it in the test cases and found that this usage, and the  `<star/>` did not save it's value when you did a `<srai>` that changed the STAR values by doing  recursive `<srai>` operations.  Dave was working on that bug and has done a fair amount of work trying to find a way to fix it.  I believe the two are related, and may be the same bug.  I think he was trying to find a sort of key to add to the STAR variable name that would be different for each level of recursion so that as your returned from the recursive operation each level would have it's own set of "star" values.

The problem only occurs if you do not first save any "star" values into permanent predicates before doing a recursive `<srai>` which goes to another `<pattern>` that uses a "*" in it's match.

The problem shows up if you try to use the original value of `<star>` in a category that did a `<srai>` to another `<pattern>`  that used the `<star>`  where the value was changed.  On return to the original upper level category, the value of star now contains the values of lower level `<srai>` patterns if there was a changed.

A work around to the problem isn't possible in AIML because once the problem occurs, the voting system gets messed up and matches do not follow the expected results causing wrong matches, or no matches in program-o.

The usage of the STAR in the NAME field is an extension of the usage of STAR added in program-o, but using STAR in a `<pattern>` is not.  The usage of STAR to catch a reduction using a `<srai>` needs to be recursive as there is no restriction of when the value of STAR can be used in the recursive operation.  The failure of the voting system breaks the "reduction" process even if the STAR value isn't needed during the recursive reduction process.

Dave has a much better understanding of why the problem exists than I do but has been unable to figure out a way to solve the issue when I last chatted with him a few days ago.

**Comment by tringate** [comment]
I have confirmed this issue is resolved.  I also have rewritten the test case as it was flawed in how it was testing this function.  It found the bug by accident I believe.

**Comment by Dave-Morton** [comment]
I thought that it may be good to point out that this bug (as well as others) wasn't the _cause_ of my need to step back, but it _was_ a **_trigger_**. I've spent literally dozens of hours on this bug alone, and it was making me insane, and has shaken my confidence a little. Just sayin' 😄

**Comment by patrickschur** [comment]
@Dave-Morton I know you stepped back and I don't want to hold you up. But can you describe to me or us how the condition in form of `<condition name="*">` works? Maybe the explaination from @tringate is already fine, so you don't have to explain it further. You would help me a lot with this. :)

**Comment by tringate** [comment]
@patrickschur, I didn't know that you were still working on this bug.  I don't think the new bug I opened, #395 is caused by using STAR.  The test case does not use STAR at all.  

I chatted some with Dave and learned that the usage of STAR in the name=  also applies to the value=.  My understanding is that the extension was to allow the usage of STAR and STARn where "n" is the index number of the star and would be valid up to the max number of STAR memory set in program-o.  I see the code that implements this in the get, set and condition tags, and suspect it is also in the srai tag as well.  I have seldom used AIML using that extension after learning of the feature.  I have not done much testing with it other than the few test cases.  After your two fixes the test cases that were failing now work, but condition tags that were working stopped working.  Only one of the three types of conditions were affected it appears.  It was the condition that used both the name= and the value= within the condition 

I am doing a bunch of work on the test cases so that errors will be much simpler to see when running the entire suite of test cases.  Each test case will be documented in the AIML file as to what it is testing and the logic used. This will be in a header comment for each test case.  My goal is to have each test case produce one of two responses.  OK for good and *** FAILED *** for bad.  I am removing the description and changing the actual test case name to all be exactly the same number of characters.  That way when run the test cases as a regression test the listing will be easy to look at and find any test case that failed.  The present structure of the output is confusing and very busy with information that is not need to determine if the regression test run was good or bad.

I also want each test case to run by itself for debugging.

**Comment by tringate** [comment]
Patrick, I moved this back to in progress based on your comments and the test results of the current test case AIML testing

**Comment by tringate** [comment]
Patrick,

I know you aqsked me is the current test case file of AIML was giving problems and which problems was it showing.  I know I answered you but it had to be in an email or something as it is not here in the bug report.

After your fixes of April 4th these are the only errors now being produced by the latest file I put in slack of the test cases AIML.

test case 023 - OK 
test case 024 - OK 
test case 025 - OK 
test case 026 - OK 
test case 027 - *** FAILED FALSE MATCH *** testing (condition) using (srai) with (star/) to undefined NAME 
test case 028 - *** FAILED FALSE MATCH *** testing (condition [NAME]) using (srai) with (star/) to undefined NAME 
test case 029 - OK 
test case 030 - *** FAILED FALSE MATCH *** testing (condition) using name="*" 
test case 031 - *** FAILED FALSE MATCH *** testing (condition [NAME]) using name="*" 
test case 032 - OK 
test case 033 - OK 
test case 034 - OK 
test case 035 - OK 

This is a rewrite of all of the test cases and they are in a different order than previous versions of that file.  They function the same but give different responses now which makes it much simpler to run and review.

The run shows the same failure but testing for two different things.  27 & 28 are looking for an "undefined" variable, 30 & 31 are using the name with an asterisk.   In both cases the value of STAR is being used, so it must have something to do with STAR.  

The form of CONDITION using name= and value= apparently works using STAR.  However, carefully looking at that AIML suggests that name=STAR may not actually be working correctly yet the logic of both test case 29 & 32 may still be failing yet producing the expected results of "ok".  The fact test case #32 shows good does make me thing it might be correctly giving a good response for both test cases, but I wanted to let you know it is possible in my mind all three might be having an issue with the STAR value given the logic I used writing the AIML test.  I believe the logic is sound, but a second set of eyes on it might not be a bad idea.
